paddle/math/float16.h

```diff
@@ -0,0 +1,714 @@
+
+#ifdef PADDLE_CUDA_FP16
+  PADDLE_HOSTDEVICE inline float16(const half& h) {
+#if CUDA_VERSION >= 9000
+    x = reinterpret_cast<__half_raw*>(&h)->x;
+#else
+    x = h.x;
+#endif  // CUDA_VERSION >= 9000
+  }
+#endif  // PADDLE_CUDA_FP16
+
+  PADDLE_HOSTDEVICE inline float16(const Eigen::half& h) : x(h.x) {}
+
+#if defined(PADDLE_NEON) && defined(PADDLE_ARM_FP16) && \
+    (PADDLE_GNUC_VER >= 61 || PADDLE_CLANG_VER >= 34)
+  // __fp16 is a native half precision data type for arm cpu,
+  // float16_t is an alias for __fp16 in arm_fp16.h,
+  // which is included in arm_neon.h.
+  PADDLE_HOSTDEVICE inline float16(const float16_t& h) {
+    float16_t tmp = h;
+    x = *reinterpret_cast<uint16_t*>(&tmp);
+  }
+#endif
+
+  PADDLE_HOSTDEVICE inline explicit float16(bool b) : x(b ? 0x3c00 : 0) {}
+
+  PADDLE_HOSTDEVICE inline explicit float16(int8_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(uint8_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(int16_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(uint16_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(int32_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(uint32_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(int64_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(uint64_t val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(float val) {
+    float16 res = fp16_impl::float_to_half_rn(val);
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline explicit float16(double val) {
+    float16 res = fp16_impl::float_to_half_rn(static_cast<float>(val));
+    x = res.x;
+  }
+
+  PADDLE_HOSTDEVICE inline float16& operator=(const float16& rhs) {
+    x = rhs.x;
+    return *this;
+  }
+
+#ifdef PADDLE_CUDA_FP16
+  PADDLE_HOSTDEVICE inline float16& operator=(const half& rhs) {
+#if CUDA_VERSION >= 9000
+    x = reinterpret_cast<__half_raw*>(&rhs)->x;
+#else
+    x = rhs.x;
+#endif
+    return *this;
+  }
+#endif
+
+  PADDLE_HOSTDEVICE inline float16& operator=(const Eigen::half& rhs) {
+    x = rhs.x;
+    return *this;
+  }
+
+#if defined(PADDLE_NEON) && defined(PADDLE_ARM_FP16) && \
+    (PADDLE_GNUC_VER >= 61 || PADDLE_CLANG_VER >= 34)
+  PADDLE_HOSTDEVICE inline float16& operator=(const float16_t& rhs) {
```

Comment:
I noticed that the following pattern

```c++
#if defined(PADDLE_NEON) && defined(PADDLE_ARM_FP16) && \
    (PADDLE_GNUC_VER >= 61 || PADDLE_CLANG_VER >= 34)
```

appeared in this file for three times. Should we define a new macro to improve the readability?

```c++
#if defined(PADDLE_NEON) && defined(PADDLE_ARM_FP16) && \
    (PADDLE_GNUC_VER >= 61 || PADDLE_CLANG_VER >= 34)
#  define PADDLE_WITH_NATIVE_FP16 
#endif
```

Reply:
Thanks! Will do.